File: web/src/features/submission-workflow/model/submission-workflow.ts
export type SubmissionPurpose = 'prior_authorization_request' | 'corrected_resubmission' | 'clinical_appeal';
export interface PacketLetter {
  readonly id: string;
  readonly purpose: SubmissionPurpose;
  readonly status: string;
  readonly version: number;
  readonly revision: number;
  readonly contentSha256Text: string;
  readonly signedAt: string | null;
}
export interface PacketAttachment {
  readonly ordinal: number;
  readonly kind: 'letter' | 'source';
  readonly name: string;
  readonly documentId: string | null;
  readonly pageCount: number;
  readonly contentSha256Text: string;
}
export interface Submission {
  readonly id: string;
  readonly commandId: string;
  readonly letterId: string;
  readonly status: string;
  readonly channel: string;
  readonly attempt: number;
  readonly submittedAt: string;
  readonly totalPages: number;
  readonly manifestSha256Text: string;
}
export interface SubmissionPacket {
  readonly caseId: string;
  readonly letter: PacketLetter | null;
  readonly submission: Submission | null;
  readonly attachments: readonly PacketAttachment[];
  readonly canSubmit: boolean;
  readonly blockReason: string | null;
}
export interface PayerReceipt {
  readonly id: string;
  readonly commandId: string;
  readonly submissionId: string;
  readonly payerReference: string;
  readonly acknowledgedAt: string;
  readonly pageCount: number;
  readonly recordedAt: string;
}
export interface ReceiptView {
  readonly packet: SubmissionPacket;
  readonly receipt: PayerReceipt | null;
  readonly custody: readonly { readonly sequence: number; readonly event: string; readonly occurredAt: string; readonly actorLabel: string }[];
}
export interface SubmitPacketCommand { readonly commandId: string; readonly expectedLetterRevision: number }
export interface AcknowledgePacketCommand {
  readonly commandId: string;
  readonly submissionId: string;
  readonly payerReference: string;
  readonly acknowledgedAt: string;
  readonly pageCount: number;
}
export type PendingSubmissionCommand =
  | { readonly kind: 'submit'; readonly body: SubmitPacketCommand }
  | { readonly kind: 'acknowledge'; readonly body: AcknowledgePacketCommand };
export interface AcknowledgementDraft {
  readonly payerReference: string;
  readonly acknowledgedAt: string;
  readonly pageCount: string;
}
export const EMPTY_ACKNOWLEDGEMENT: AcknowledgementDraft = { payerReference: '', acknowledgedAt: '', pageCount: '' };
export function signedCurrentPacket(packet: SubmissionPacket): boolean {
  return packet.canSubmit && packet.letter?.status === 'signed' && packet.letter.signedAt !== null;
}
export function acknowledgementBody(draft: AcknowledgementDraft, submissionId: string, commandId: string): AcknowledgePacketCommand | null {
  const pageCount = Number(draft.pageCount);
  const date = new Date(draft.acknowledgedAt);
  if (!draft.payerReference.trim() || !draft.pageCount.trim() || !Number.isSafeInteger(pageCount) || pageCount < 0 || !Number.isFinite(date.getTime())) return null;
  return { commandId, submissionId, payerReference: draft.payerReference.trim(), acknowledgedAt: date.toISOString(), pageCount };
}
function record(value: unknown): value is Record<string, unknown> { return typeof value === 'object' && value !== null && !Array.isArray(value); }
function fields(value: unknown, keys: readonly string[]): value is Record<string, unknown> { return record(value) && keys.every((key) => typeof value[key] === 'string'); }
function integer(value: unknown, minimum = 0): boolean { return Number.isSafeInteger(value) && (value as number) >= minimum; }
export function parsePacket(value: unknown, caseId: string): SubmissionPacket {
  if (!record(value) || value.caseId !== caseId || typeof value.canSubmit !== 'boolean' || !(value.blockReason === null || typeof value.blockReason === 'string') || !Array.isArray(value.attachments)) throw new Error('Invalid submission packet.');
  const letter = value.letter;
  if (letter !== null && (!fields(letter, ['id', 'purpose', 'status', 'contentSha256Text']) || !['prior_authorization_request', 'corrected_resubmission', 'clinical_appeal'].includes(letter.purpose as string) || !integer(letter.version, 1) || !integer(letter.revision, 1) || !(letter.signedAt === null || typeof letter.signedAt === 'string'))) throw new Error('Invalid packet letter.');
  const submission = value.submission;
  if (submission !== null && (!fields(submission, ['id', 'commandId', 'letterId', 'status', 'channel', 'submittedAt', 'manifestSha256Text']) || !integer(submission.attempt, 1) || !integer(submission.totalPages, 1))) throw new Error('Invalid packet transmission.');
  if (!value.attachments.every((item) => fields(item, ['kind', 'name', 'contentSha256Text']) && ['letter', 'source'].includes(item.kind as string) && integer(item.ordinal, 1) && integer(item.pageCount, 1) && (item.documentId === null || typeof item.documentId === 'string'))) throw new Error('Invalid attachment manifest.');
  return value as unknown as SubmissionPacket;
}
export function parseReceiptView(value: unknown, caseId: string): ReceiptView {
  if (!record(value) || !Array.isArray(value.custody)) throw new Error('Invalid receipt view.');
  const packet = parsePacket(value.packet, caseId);
  const receipt = value.receipt;
  if (receipt !== null && (!fields(receipt, ['id', 'commandId', 'submissionId', 'payerReference', 'acknowledgedAt', 'recordedAt']) || !integer(receipt.pageCount) || receipt.submissionId !== packet.submission?.id)) throw new Error('Receipt does not match this submission.');
  if (!value.custody.every((item) => fields(item, ['event', 'occurredAt', 'actorLabel']) && integer(item.sequence, 1))) throw new Error('Invalid custody record.');
  return { packet, receipt: receipt as PayerReceipt | null, custody: value.custody as ReceiptView['custody'] };
}

File: crates/aso-host/src/submission_workflow.rs
//! Durable submission packet, acknowledgement, and custody contracts.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::{
    AppServices, affirmation::ClinicalContext, letter_workflow::LetterPurpose, session::Principal,
};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct SubmitPacketCommand {
    pub command_id: Uuid,
    pub expected_letter_revision: i64,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct RecordAcknowledgementCommand {
    pub command_id: Uuid,
    pub submission_id: Uuid,
    pub payer_reference: String,
    pub acknowledged_at: DateTime<Utc>,
    pub page_count: i32,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct SubmissionLetterSnapshot {
    pub id: Uuid,
    pub purpose: LetterPurpose,
    pub status: String,
    pub version: i32,
    pub revision: i64,
    pub content_sha256_text: String,
    pub signed_at: Option<DateTime<Utc>>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct SubmissionAttachmentSnapshot {
    pub ordinal: i32,
    pub kind: String,
    pub name: String,
    pub document_id: Option<Uuid>,
    pub page_count: i32,
    pub content_sha256_text: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct SubmissionSnapshot {
    pub id: Uuid,
    pub command_id: Uuid,
    pub letter_id: Uuid,
    pub status: String,
    pub channel: String,
    pub attempt: i32,
    pub submitted_at: DateTime<Utc>,
    pub total_pages: i32,
    pub manifest_sha256_text: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct SubmissionPacketSnapshot {
    pub case_id: Uuid,
    pub letter: Option<SubmissionLetterSnapshot>,
    pub submission: Option<SubmissionSnapshot>,
    pub attachments: Vec<SubmissionAttachmentSnapshot>,
    pub can_submit: bool,
    pub block_reason: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct ReceiptSnapshot {
    pub id: Uuid,
    pub command_id: Uuid,
    pub submission_id: Uuid,
    pub payer_reference: String,
    pub acknowledged_at: DateTime<Utc>,
    pub page_count: i32,
    pub recorded_at: DateTime<Utc>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct CustodyEventSnapshot {
    pub sequence: i32,
    pub event: String,
    pub occurred_at: DateTime<Utc>,
    pub actor_label: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct SubmissionReceiptView {
    pub packet: SubmissionPacketSnapshot,
    pub receipt: Option<ReceiptSnapshot>,
    pub custody: Vec<CustodyEventSnapshot>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, thiserror::Error)]
pub enum SubmissionWorkflowError {
    #[error("authentication required")]
    Unauthenticated,
    #[error("submission workflow denied")]
    Denied,
    #[error("case, letter, submission, or command not found")]
    NotFound,
    #[error("the packet changed after it was loaded")]
    RevisionConflict,
    #[error("command identifier already has a different payload")]
    CommandConflict,
    #[error("the packet is not ready for submission")]
    NotReady,
    #[error("submission request is invalid")]
    Invalid,
    #[error("submission service unavailable")]
    Unavailable,
}

impl AppServices {
    fn check_submission_context(
        &self,
        context: &ClinicalContext,
        capabilities: &[String],
        require_submit: bool,
    ) -> Result<(), SubmissionWorkflowError> {
        if context.expires_at <= self.clock.now() {
            return Err(SubmissionWorkflowError::Unauthenticated);
        }
        if context.principal != Principal::User {
            return Err(SubmissionWorkflowError::Denied);
        }
        let allowed = capabilities
            .iter()
            .any(|value| value == "submit" || (!require_submit && value == "case:read"));
        allowed.then_some(()).ok_or(SubmissionWorkflowError::Denied)
    }

    pub async fn read_submission_packet(
        &self,
        context: &ClinicalContext,
        capabilities: &[String],
        case_id: Uuid,
    ) -> Result<SubmissionPacketSnapshot, SubmissionWorkflowError> {
        self.check_submission_context(context, capabilities, false)?;
        if case_id.is_nil() {
            return Err(SubmissionWorkflowError::Invalid);
        }
        self.letters.read_submission_packet(context, case_id).await
    }

    pub async fn submit_packet(
        &self,
        context: &ClinicalContext,
        capabilities: &[String],
        case_id: Uuid,
        command: &SubmitPacketCommand,
    ) -> Result<SubmissionPacketSnapshot, SubmissionWorkflowError> {
        self.check_submission_context(context, capabilities, true)?;
        if case_id.is_nil() || command.command_id.is_nil() || command.expected_letter_revision < 1 {
            return Err(SubmissionWorkflowError::Invalid);
        }
        self.letters.submit_packet(context, case_id, command).await
    }

    pub async fn read_submission_receipt(
        &self,
        context: &ClinicalContext,
        capabilities: &[String],
        case_id: Uuid,
    ) -> Result<SubmissionReceiptView, SubmissionWorkflowError> {
        self.check_submission_context(context, capabilities, false)?;
        if case_id.is_nil() {
            return Err(SubmissionWorkflowError::Invalid);
        }
        self.letters.read_submission_receipt(context, case_id).await
    }

    pub async fn record_submission_acknowledgement(
        &self,
        context: &ClinicalContext,
        capabilities: &[String],
        case_id: Uuid,
        command: &RecordAcknowledgementCommand,
    ) -> Result<SubmissionReceiptView, SubmissionWorkflowError> {
        self.check_submission_context(context, capabilities, true)?;
        if case_id.is_nil()
            || command.command_id.is_nil()
            || command.submission_id.is_nil()
            || command.payer_reference.trim().is_empty()
            || command.page_count < 0
        {
            return Err(SubmissionWorkflowError::Invalid);
        }
        self.letters
            .record_submission_acknowledgement(context, case_id, command)
            .await
    }
}
